CSharp/Library/Microsoft.Bot.Builder/Fibers/Awaitable.cs

```diff
@@ -95,6 +96,55 @@ IAwaiter<T> IAwaitable<T>.GetAwaiter()
         }
     }
 
+    /// <summary>
+    /// Creates a <see cref="IAwaitable{T}"/> from source passed to constructor.
+    /// </summary>
+    /// <typeparam name="TSource"> The type of the source for build the item.</typeparam>
+    /// <typeparam name="TItem">The type if the item.</typeparam>
+    public sealed class AwaitableFromSource<TSource, TItem> : IAwaitable<TItem>, IAwaiter<TItem>
+    {
+        private readonly TSource source;
+        private readonly Func<TSource, Task<TItem>> resolver;
+
+        private bool resolved = false;
+        private TItem result;
+
+        public AwaitableFromSource(TSource source, Func<TSource, Task<TItem>> resolver)
+        {
+            this.source = source;
+            this.resolver = resolver;
+        }
+
+        public bool IsCompleted
+        {
+            get
+            {
+                return this.resolved;
+            }
+        }
+
+        public IAwaiter<TItem> GetAwaiter()
+        {
+            return this;
+        }
+
+        public TItem GetResult()
+        {
+            return this.result;
+        }
+
+        public async void OnCompleted(Action continuation)
```

Comment:
We still need to do the async part on another thread or otherwise a deadlock will happen:

`this.result = Task.Run(() => this.resolver(this.source)).Result;`